---
title: "Predict Ethnicity"
output: 
    rmarkdown::html_vignette:
        fig_caption: FALSE
        toc: TRUE
pkgdown:
    as_is: true
vignette: >
    %\VignetteIndexEntry{Predict Ethnicity}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---



### Installation


```r
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("planet")
```

### Example Data

For demonstration, I use 24 samples from a placental DNAm dataset from GEO,
([GSE7519](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75196)), which
contains samples collected in an Australian population. The DNA methylation 
data (in betas) can be accessed with `data(plBetas)` and corresponding sample
information from `data(plPhenoData)`. Note that cpgs have been filtered to a 
random \~10,000 CpGs, plus the CpGs used inall of the models from this package.


```r
library(dplyr) # for data wrangling
library(ggplot2) #  plotting
library(planet)

# load example data
data(plBetas)
data(plPhenoData)

dim(plBetas)
#> [1] 13918    24
head(plPhenoData)
#> # A tibble: 6 x 7
#>   sample_id  sex   disease   gestation_wk ga_RPC ga_CPC ga_RRPC
#>   <fct>      <chr> <chr>            <dbl>  <dbl>  <dbl>   <dbl>
#> 1 GSM1944936 Male  preeclam~           36   38.5   38.7    38.7
#> 2 GSM1944939 Male  preeclam~           32   33.1   34.2    32.6
#> 3 GSM1944942 Fema~ preeclam~           32   34.3   35.1    33.3
#> 4 GSM1944944 Male  preeclam~           35   35.5   36.7    35.5
#> 5 GSM1944946 Fema~ preeclam~           38   37.6   37.6    36.6
#> 6 GSM1944948 Fema~ preeclam~           36   36.8   38.4    36.7
```

### Infer Ethnicity

**Requirements:**

- CpGs and SNPs that are used in `pl_infer_ethnicity` need to be in your betas
`data.frame`. You can ensure that you have all features with the
`ethnicityCpGs` vector:


```r
data(ethnicityCpGs)
all(ethnicityCpGs %in% rownames(plBetas))
#> [1] TRUE
```

*Note that if some features are missing, `pl_infer_ethnicity` will throw a
warning, but will still work.*

- The betas `data.frame` needs to have *samples in columns* and *CpGs/snps* in
rows. The rownames must be CpG/rs identifiers.

- If you have IDAT files available, then I recommend normalizing your betas
`data.frame` using the same normalization methods used on the training data:
[**noob**](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3627582/) and
[**BMIQ**](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3546795/). To apply 
these, run `minfi::preprocessNoob()` on an `rgset` object and then
`wateRmelon::BMIQ()`. This has already applied to the example data.


```r
results <- predictEthnicity(plBetas)

# show last 8 rows
results %>%
    tail(8)
#> # A tibble: 8 x 7
#>   Sample_ID Predicted_ethni~ Predicted_ethni~ Prob_African
#>   <chr>     <chr>            <chr>                   <dbl>
#> 1 GSM19449~ Asian            Asian                0.0123  
#> 2 GSM19449~ Caucasian        Caucasian            0.0157  
#> 3 GSM19449~ Asian            Asian                0.0208  
#> 4 GSM19449~ Caucasian        Caucasian            0.000928
#> 5 GSM19449~ Caucasian        Caucasian            0.00226 
#> 6 GSM19449~ Caucasian        Caucasian            0.00660 
#> 7 GSM19449~ Caucasian        Caucasian            0.00216 
#> 8 GSM19449~ Caucasian        Caucasian            0.00114 
#> # ... with 3 more variables: Prob_Asian <dbl>,
#> #   Prob_Caucasian <dbl>, Highest_Prob <dbl>
```

`predictEthnicity` returns probabilities corresponding to each ethnicity for
each sample (e.g `Prob_Caucasian`, `Prob_African`, `Prob_Asian`). A final
classification is determined in two ways:

1.  `Predicted_ethnicity_nothresh` - returns a classification corresponding to
the highest class-specific probability.

2.  `Predicted_ethnicity` - if the highest class-specific probability is below
`0.75`, then the the sample is assigned an `Amibiguous` label. This threshold 
can be adjusted with the `threshold` argument. Samples with this label might
require special attention in downstream analyses.


```r
results %>%
    ggplot(aes(
        x = Prob_Caucasian, y = Prob_African,
        col = Predicted_ethnicity
    )) +
    geom_point(alpha = 0.7) +
    coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
    scale_x_continuous(labels = scales::percent) +
    scale_y_continuous(labels = scales::percent) +
    labs(x = "P(Caucasian)", y = "P(African)")
```

<img src="fig/ethnicity-1.png" title="plot of chunk ethnicity" alt="plot of chunk ethnicity" style="display: block; margin: auto;" />

```r

results %>%
    ggplot(aes(
        x = Prob_Caucasian, y = Prob_Asian,
        col = Predicted_ethnicity
    )) +
    geom_point(alpha = 0.7) +
    coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
    scale_x_continuous(labels = scales::percent) +
    scale_y_continuous(labels = scales::percent) +
    labs(x = "P(Caucasian)", y = "P(Asian)")
```

<img src="fig/ethnicity-2.png" title="plot of chunk ethnicity" alt="plot of chunk ethnicity" style="display: block; margin: auto;" />

We can't compare this to self-reported ethnicity as it is unavailable. But we
know these samples were collected in Sydney, Australia, and are therefore 
likely mostly European with some East Asian ancestries.


```r
table(results$Predicted_ethnicity)
#> 
#>     Asian Caucasian 
#>         2        22
```

**A note on adjustment in differential methylation analysis**

Because 'Ambiguous' samples might have different mixtures of ancestries, it 
might be inadequate to adjust for them as one group in an analysis of admixed
populations (e.g. 50/50 Asian/African should not be considered the same group 
as 50/50 Caucasian/African). One solution would be to simply remove these
samples. Another would be to adjust for the raw probabilities-in this case, use
only two of the three probabilities, since the third will be redundant
(probabilities sum to 1). If sample numbers are large enough in each group,
stratifying downstream analyses by ethnicity might also be a valid option.

### References

1. [**Yuan V**, Price EM, Del Gobbo G, Mostafavi S, Cox B, Binder AM, et al.
Accurate ethnicity prediction from placental DNA methylation data. Epigenetics 
& Chromatin. 2019 Aug 9;12(1):51.](https://epigeneticsandchromatin.
biomedcentral.com/articles/10.1186/s13072-019-0296-3)

### Session Info


```r
sessionInfo()
#> R Under development (unstable) (2020-12-18 r79646)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19042)
#> 
#> Matrix products: default
#> 
#> locale:
#> [1] LC_COLLATE=English_Canada.1252 
#> [2] LC_CTYPE=English_Canada.1252   
#> [3] LC_MONETARY=English_Canada.1252
#> [4] LC_NUMERIC=C                   
#> [5] LC_TIME=English_Canada.1252    
#> 
#> attached base packages:
#> [1] stats4    parallel  stats     graphics  grDevices utils    
#> [7] datasets  methods   base     
#> 
#> other attached packages:
#>  [1] planet_0.99.1               tidyr_1.1.2                
#>  [3] dplyr_1.0.2                 EpiDISH_2.7.0              
#>  [5] bumphunter_1.33.0           locfit_1.5-9.4             
#>  [7] iterators_1.0.13            foreach_1.5.1              
#>  [9] Biostrings_2.59.2           XVector_0.31.1             
#> [11] SummarizedExperiment_1.21.1 Biobase_2.51.0             
#> [13] MatrixGenerics_1.3.0        matrixStats_0.57.0         
#> [15] GenomicRanges_1.43.1        GenomeInfoDb_1.27.3        
#> [17] IRanges_2.25.6              S4Vectors_0.29.6           
#> [19] BiocGenerics_0.37.0         ggplot2_3.3.3              
#> 
#> loaded via a namespace (and not attached):
#>  [1] nlme_3.1-151              bitops_1.0-6             
#>  [3] bit64_4.0.5               filelock_1.0.2           
#>  [5] progress_1.2.2            httr_1.4.2               
#>  [7] rprojroot_2.0.2           tools_4.1.0              
#>  [9] doRNG_1.8.2               utf8_1.1.4               
#> [11] R6_2.5.0                  mgcv_1.8-33              
#> [13] DBI_1.1.0                 colorspace_2.0-0         
#> [15] withr_2.3.0               tidyselect_1.1.0         
#> [17] prettyunits_1.1.1         bit_4.0.4                
#> [19] curl_4.3                  compiler_4.1.0           
#> [21] cli_2.2.0                 xml2_1.3.2               
#> [23] DelayedArray_0.17.7       labeling_0.4.2           
#> [25] rtracklayer_1.51.3        scales_1.1.1             
#> [27] quadprog_1.5-8            askpass_1.1              
#> [29] rappdirs_0.3.1            stringr_1.4.0            
#> [31] digest_0.6.27             Rsamtools_2.7.0          
#> [33] rmarkdown_2.6             htmltools_0.5.0          
#> [35] pkgconfig_2.0.3           sparseMatrixStats_1.3.2  
#> [37] highr_0.8                 dbplyr_2.0.0             
#> [39] rlang_0.4.9               rstudioapi_0.13          
#> [41] RSQLite_2.2.1             DelayedMatrixStats_1.13.2
#> [43] farver_2.0.3              BiocIO_1.1.2             
#> [45] generics_0.1.0            BiocParallel_1.25.2      
#> [47] RCurl_1.98-1.2            magrittr_2.0.1           
#> [49] GenomeInfoDbData_1.2.4    Matrix_1.3-2             
#> [51] fansi_0.4.1               Rcpp_1.0.5               
#> [53] munsell_0.5.0             lifecycle_0.2.0          
#> [55] stringi_1.5.3             yaml_2.2.1               
#> [57] MASS_7.3-53               zlibbioc_1.37.0          
#> [59] BiocFileCache_1.15.1      grid_4.1.0               
#> [61] blob_1.2.1                crayon_1.3.4             
#> [63] lattice_0.20-41           splines_4.1.0            
#> [65] GenomicFeatures_1.43.3    hms_0.5.3                
#> [67] knitr_1.30                pillar_1.4.7             
#> [69] rjson_0.2.20              rngtools_1.5             
#> [71] codetools_0.2-18          biomaRt_2.47.1           
#> [73] XML_3.99-0.5              glue_1.4.2               
#> [75] evaluate_0.14             vctrs_0.3.6              
#> [77] locfdr_1.1-8              gtable_0.3.0             
#> [79] openssl_1.4.3             purrr_0.3.4              
#> [81] assertthat_0.2.1          xfun_0.19                
#> [83] restfulr_0.0.13           e1071_1.7-4              
#> [85] class_7.3-17              tibble_3.0.4             
#> [87] GenomicAlignments_1.27.2  AnnotationDbi_1.53.0     
#> [89] memoise_1.1.0             ellipsis_0.3.1           
#> [91] here_1.0.1
```
